webserver.py

```diff
@@ -41,61 +42,178 @@ def get_conn():
-            if game_state.get('plot' + str(i) + '_' + str(j)) is None:
-                game_state['plot' + str(i) + '_' + str(j)] = {'seedType': 0, 'sowTime': 0, 'locked': 1}
-            if get_plot(game_state, i, j)['seedType'] != 0:
-                total_crops_planted += 1
-    for i in range(2):
-        for j in range(2):
-            get_plot(game_state, i, j)['locked'] = 0
-    if total_seed_count == 0 and total_crops_planted == 0:
-        if game_state['resources']['cash'] < GAME_CONFIG['starting_resources']['cash']:
-            game_state['resources']['cash'] = GAME_CONFIG['starting_resources']['cash']
-    if game_state.get('recipes') is None:
-        game_state['recipes'] = []
-    return game_state
-
-
-def save_state(slug, game_state):
-    my_conn = get_conn()
-    json_game_state = json.dumps(game_state, indent=4)
-    with my_conn:
-        with my_conn.cursor() as cursor:
-            cursor.execute('INSERT INTO game_states (slug, game_state) VALUES (%s, %s) '
-                           'ON CONFLICT (slug) DO UPDATE SET game_state=%s;',
-                           [slug, json_game_state, json_game_state])
+class GameState(object):
+    def __init__(self, json_data):
+        self.data = json_data
+
+    @staticmethod
+    def new(slug, password):
+        game_state = GameState.load(slug, password)
+        if game_state:
+            raise Forbidden("Username already taken")
+        else:
+            game_state = GameState({
+                'resources': {},
+                'slug': slug,
+                'password': password,
+                'unlockCount': 0,
+                'seedCounts': {},
+                'recipes': ['a']
+            })
+        return game_state.initialize()
+
+    @classmethod
+    def load(cls, slug, password):
+        my_conn = get_conn()
+        with my_conn:
+            with my_conn.cursor() as cursor:
+                cursor.execute('SELECT game_state FROM game_states WHERE slug = %s;', [slug])
+                row = cursor.fetchone()
+                if not row:
+                    return None
+                game_state = GameState(row[0])
+                game_state.check_password(password)
+                return game_state.initialize()
+
+    def save(self, slug):
+        my_conn = get_conn()
+        json_game_state = json.dumps(self.data, indent=4)
+        with my_conn:
+            with my_conn.cursor() as cursor:
+                cursor.execute('INSERT INTO game_states (slug, game_state) VALUES (%s, %s) '
+                               'ON CONFLICT (slug) DO UPDATE SET game_state=%s;',
+                               [slug, json_game_state, json_game_state])
+
+    # add missing seed counts, resources, and plots to game_state
+    # unlock first four plots
+    # check to never run out of money
+    def initialize(self):
+        for seed_id in GAME_CONFIG['seeds']:
+            if self.data['seedCounts'].get(seed_id) is None:
```

Comment:
Now that we've got a class, this can be cleaned up a lot to make it more readable.

Rather than `if self.data['seedCounts'].get(seed_id) is None...` checks, add a `get_seed_count(seed_id)` method that returns 0 if the seed's not found.  Then this code becomes much cleaner.